File: Grind_75/minStack.js
var MinStack = function() {
    this.stack = [];
    this.min = [];
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    if (!this.min.length || val <= this.min[this.min.length - 1]) {
        this.min.push(val)
    }
    this.stack.push(val)

   return this.stack;
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    let popped =  this.stack.pop();
    if (popped === this.min[this.min.length - 1]) this.min.pop()
    return popped;
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.stack[this.stack.length - 1];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.min[this.min.length - 1];
};

const minStack = new MinStack();
minStack.push(20);
minStack.push(15);
minStack.push(12);
minStack.pop();
minStack.pop();
console.log(minStack.getMin());

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: PythonStuff/minStack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.min = []

    def push(self, val: int) -> None:
        try:
            if val < self.min[len(self.min) - 1]:
                self.min.append(val)
        except:
            self.min.append(val)

        self.stack.append(val)

    def pop(self) -> None:
        popped = self.stack.pop()

        if popped == self.min[len(self.min) - 1]:
            self.min.pop()

        return popped

    def top(self) -> int:
        return self.stack[len(self.stack) - 1]

    def getMin(self) -> int:
        return self.min[len(self.min) - 1]

minStack = MinStack()
minStack.push(20)
minStack.push(15)
minStack.push(12)
minStack.pop()
minStack.pop()
print(minStack.getMin())